src/libraries/System.Private.CoreLib/src/System/Single.cs

```diff
@@ -849,10 +849,12 @@ bool IFloatingPoint<float>.TryWriteSignificandLittleEndian(Span<byte> destinatio
 
         /// <inheritdoc cref="IFloatingPointIeee754{TSelf}.ReciprocalEstimate(TSelf)" />
         [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
         public static float ReciprocalEstimate(float x) => MathF.ReciprocalEstimate(x);
 
         /// <inheritdoc cref="IFloatingPointIeee754{TSelf}.ReciprocalSqrtEstimate(TSelf)" />
         [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
         public static float ReciprocalSqrtEstimate(float x) => MathF.ReciprocalSqrtEstimate(x);
```

Comment:
There ended up being 2 issues at play.

1. The APIs weren't recursive and so the delegate invocation returned a different result. The importercalls.cpp changes fix that.
2. There's some R2R quirk related to S.P.Corelib that I've not entirely figured out yet

For 2, the rules for how `compExactlyDependsOn` is meant to work is laid out here: https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/vectors-and-intrinsics.md. While there is some in depth description of `CompExactlyDependsOnAttribute` and how it needs to be distributed on the method and all callers of the method within S.P.Corelib, it doesn't relay that same piece of information about the JIT's `compExactlyDependsOn` check, which should be getting done for any API calling the intrinsic already, thus marking it as requiring `AVX512F` for example.

I believe the issue ends up being similar to what's coded in https://github.com/dotnet/runtime/blob/main/src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs#L593, which is that if at least one of the types is actually supported (such as `SSE`) then it will be treated as supported for R2R purposes. Thus, you cannot mark something as `CompExactlyDependsOn(typeof(Avx512F))` and expect R2R to be skipped (at least for corelib).

Thus, I've explicitly marked the methods as `AggressiveOptimization` instead to prevent them from being used in R2R at all to workaround the issue.

Reply:
CC. @jkotas, @davidwrighton 

This should fix the test failure, but I'd appreciate some insight here as to whether there is a better way to get these methods always excluded from R2R, since they are intentionally non-deterministic across different hardware for performance purposes.

Reply:
I do not think the current model has a simple way to describe intrinsics methods that are non-deterministic across different hardware. You would have to model it explicitly using something like:
```
ReciprocalEstimate()
{
    if (AVX512F.IsSupported)
    {
        return ReciprocalEstimate_AVX512F();
    }
    else if (AdvSimd.IsSupported)
    {
        return ReciprocalEstimate_AdvSimd();
    }
...
}

[CompExactlyDependsOn(typeof(AVX512F))]
ReciprocalEstimate_AVX512F() => ReciprocalEstimate_AVX512F();

[CompExactlyDependsOn(typeof(AdvSimd))]
ReciprocalEstimate_AdvSimd() => ReciprocalEstimate_AdvSimd();
```
It is probably more complicated that you want.


If you just want to disable R2R for the method without additional side-effects of AggressiveOptimization, you can mark it with `System.Runtime.BypassReadyToRunAttribute`. (Define an internal copy of the attribute - it is not part of public surface.)


Do `ConvertToIntegerNative` intrinsics that you have added some time ago have the same non-deterministic across different hardware.problem?

Reply:
> If you just want to disable R2R for the method without additional side-effects of AggressiveOptimization, you can mark it with System.Runtime.BypassReadyToRunAttribute. (Define an internal copy of the attribute - it is not part of public surface.)

👍 

> Do ConvertToIntegerNative intrinsics that you have added some time ago have the same non-deterministic across different hardware problem?

Yes. They just do whatever is faster for the current hardware leaving the undefined behavior only deterministic within a single process.

So, for example, `float -> uint` has to emulate everything on SSE-AVX2 hardware; but can directly emit the singular `vcvtss2usi` intrinsic on AVX512 hardware, allowing it to be a single instruction rather than a sequence.

I'll mark them with the attribute as well as part of this.

Reply:
Things are passing locally with the latest change.

The `[BypassReadyToRun]` on the recursive functions ensures that R2R won't try to handle those APIs itself (so we don't end up with R2R having an infinitely recursing method).

The `if (opts.ReadyToRun())` checks for the normal expansion path then ensure they aren't expanded in R2R images for other callers.

I'm going to let CI run and then will update (in the AM) https://github.com/dotnet/runtime/blob/main/docs/design/coreclr/botr/vectors-and-intrinsics.md with a note about this setup for such APIs and add an explicit regression test covering the minimal repro for these APIs, which essentially looks like (adjusting `value` based on the API being tested):
```csharp
const float value = -1E-45f;
Func<float, float> action = float.ReciprocalSqrtEstimate;

float expectedResult = float.ReciprocalSqrtEstimate(value);
float actualResult = action(value);

Assert.Equal(expectedResult, actualResult);
```